contracts/StakeManager.sol

```diff
@@ -0,0 +1,165 @@
+    function stakeForAddress(address registree, uint256 unstakeDelay) external payable {
+        require(registrees[msg.sender] == address(0) || registrees[msg.sender] == registree, "different stake exists");
+        require(stakes[registree].owner == address(0) || stakes[registree].owner == msg.sender, "not owner");
+        require(unstakeDelay >= stakes[registree].unstakeDelay, "unstakeDelay cannot be decreased");
+        require(msg.sender != registree, "registree cannot stake for itself");
+        require(stakes[msg.sender].owner == address(0), "sender is a registree itself");
+        registrees[msg.sender] = registree;
+        stakes[registree].owner = msg.sender;
+        stakes[registree].stake += msg.value;
+        stakes[registree].unstakeDelay = unstakeDelay;
+        emit StakeAdded(registree, stakes[registree].owner, stakes[registree].stake, stakes[registree].unstakeDelay);
+    }
+
+    function unlockStake() external {
+        address registree = registrees[msg.sender];
+        StakeInfo storage info = stakes[registree];
+        require(registree != address(0), "not owner");
+        require(info.withdrawBlock == 0, "already pending");
+        info.withdrawBlock = block.number + info.unstakeDelay;
+        emit StakeUnlocked(registree, msg.sender, info.withdrawBlock);
+    }
+
+    function withdrawStake() external {
+        address registree = registrees[msg.sender];
+        require(registree != address(0), "not owner");
+        require(stakes[registree].withdrawBlock > 0, "Withdrawal is not scheduled");
+        require(stakes[registree].withdrawBlock <= block.number, "Withdrawal is not due");
+        uint256 amount = stakes[registree].stake;
+        delete stakes[registree];
+        delete registrees[msg.sender];
+        msg.sender.transfer(amount);
+        emit StakeWithdrawn(registree, msg.sender, amount);
+    }
+
+    function authorizeHub(address relayHub) external {
+        address registree = registrees[msg.sender];
+        require(registree != address(0), "not owner");
+        authorizedHubs[registree][relayHub].removalBlock = uint(-1);
+        emit HubAuthorized(registree, relayHub);
+    }
+
+    function unauthorizeHub(address relayHub) external {
+        address registree = registrees[msg.sender];
+        require(registree != address(0), "not owner");
+        RelayHubInfo storage hubInfo = authorizedHubs[registree][relayHub];
+        require(hubInfo.removalBlock == uint(-1), "hub not authorized");
+        uint256 removalBlock = block.number + stakes[registree].unstakeDelay;
+        hubInfo.removalBlock = removalBlock;
+        emit HubUnauthorized(registree, relayHub, removalBlock);
+    }
+
+    function isRegistreeStaked(address registree, uint256 minAmount, uint256 minUnstakeDelay)
+    external
+    view
+    returns (bool) {
+        StakeInfo storage info = stakes[registree];
+        bool isAmountSufficient = info.stake > minAmount;
```

Comment:
better have them as separate `require` items. This way they also provide some info on failure

Reply:
I am not sure I want this to revert for failure. I can provide return codes, but this is just 4 flags, it should be reasonably easy to figure out anyways

Reply:
revert is excellent, especially where this is exactly what the caller (RelayHub) would do with the result

Reply:
I don't want to make this decision for the Relay Hub. I think returning boolean does the job

Reply:
You don't make a decision here. returning a boolean is simply hiding the reason for returning false.